### PR TITLE
Add Window.onWindowFocusChangedListeners API

### PR DESCRIPTION
**Pull request by consp1racy**
Following up on https://github.com/square/curtains/issues/3

The API largely continues what `OnContentChangedListener` established.

Tests check that the callback isn't invoked immediately with a current value (I don't think we know it anyway).
Tests check that activity focus changes after being resumed and before being paused.
There isn't a test case for activity gaining focus after a dialog is dismissed. At this point we're just testing the framework instead of the library, I think.
I ran tests locally on AOSP AVDs API 19 and 21-29.

### CODE REVIEW

**Comment by consp1racy** [comment]
Ran apiDump and ktlintFormat

**Comment by pyricau** [comment]
Mmm api 16 failed with unrelated error. 

curtains.test.OnNextDrawTest > onNextDraw_triggers_on_first_activity_draw[test(AVD) - 4.1.2] FAILED 
	expected: 3
	but was : 2


Looks like a flake? I should investigate but let's re kick the build.

Other than that LGTM, let me.know if good to merge of if you're still iterating.

**Comment by pyricau** [comment]
API 30 No good:

curtains.test.OnWindowFocusChangedListenersTest > activity_focus_lost_on_dialog_showed[test(AVD) - 11] FAILED 
	Expected latch to count to zero within 30 seconds, count is still at 1
	but was: java.util.concurrent.CountDownLatch@7717e5e[Count = 1]

curtains.test.OnWindowFocusChangedListenersTest > activity_focus_gained_on_activity_resumed[test(AVD) - 11] FAILED 
	Expected latch to count to zero within 30 seconds, count is still at 1
	but was: java.util.concurrent.CountDownLatch@8a30629[Count = 1]

curtains.test.OnWindowFocusChangedListenersTest > activity_focus_lost_on_activity_paused[test(AVD) - 11] FAILED 
	Expected latch to count to zero within 30 seconds, count is still at 1
	but was: java.util.concurrent.CountDownLatch@7f64312[Count = 1]

curtains.test.OnWindowFocusChangedListenersTest > dialog_focus_gained_on_dialog_showed[test(AVD) - 11] FAILED 
	Expected latch to count to zero within 30 seconds, count is still at 1
	but was: java.util.concurrent.CountDownLatch@7717e5e[Count = 1]

**Review by pyricau — APPROVED** [review]

**Comment by pyricau** [comment]
Weird I can't repro the failure locally on a API 30 emulator.

**Comment by consp1racy** [comment]
If it's fine locally, you can go ahead and merge

**Comment by pyricau** [comment]
> If it's fine locally, you can go ahead and merge

Past experience has taught me it's usually worth digging, often there's a sneaky bug hiding :)

**Comment by consp1racy** [comment]
Hmm, probably one of these things affects the timing or there's a race condition.

> As a general rule, however, a foreground activity will have window focus... unless it has displayed other dialogs or popups that take input focus, in which case the activity itself will not have focus when the other windows have it. Likewise, the system may display system-level windows (such as the status bar notification panel or a system alert) which will temporarily take window input focus without pausing the foreground activity.

> Starting with Build.VERSION_CODES.Q there can be multiple resumed activities at the same time in multi-window mode, so resumed state does not guarantee window focus even if there are no overlays above.

> If the intent is to know when an activity is the topmost active, the one the user interacted with last among all activities but not including non-activity windows like dialogs and popups, then onTopResumedActivityChanged(boolean) should be used. On platform versions prior to Build.VERSION_CODES.Q, onResume() is the best indicator.

https://developer.android.com/reference/android/app/Activity#onWindowFocusChanged(boolean)

**Comment by consp1racy** [comment]
Functionally it should be good, but probably worth poking at later

**Comment by consp1racy** [comment]
activity_focus_lost_on_activity_paused failed for me locally on some 4th run on API 30 Pixel 4 Google APIs:

expected to be false
at curtains.test.OnWindowFocusChangedListenersTest$activity_focus_lost_on_activity_paused$$inlined$use$lambda$1$1.onWindowFocusChanged(OnWindowFocusChangedListenersTest.kt:50)
at curtains.internal.WindowCallbackWrapper.onWindowFocusChanged(WindowCallbackWrapper.kt:50)
at com.android.internal.policy.DecorView.onWindowFocusChanged(DecorView.java:1691)

**Comment by pyricau** [comment]
> Mmm api 16 failed with unrelated error.
> 
> curtains.test.OnNextDrawTest > onNextDraw_triggers_on_first_activity_draw[test(AVD) - 4.1.2] FAILED
> expected: 3
> but was : 2
> 
> Looks like a flake? I should investigate but let's re kick the build.
> 
> Other than that LGTM, let me.know if good to merge of if you're still iterating.

Fix: https://github.com/square/curtains/pull/7

**Comment by consp1racy** [comment]
Yea, I sometimes see a focus gained event before focus lost. I'm tempted to do this 😂

no focus

> I wonder if we should do the same and have OnWindowFocusGainedListener and OnWindowFocusLostListener

I like that, but before I add it, does it make sense?

In my case I also needed to track when the last focused activity lost focus. That's not possible with just focus gained listener
```
when {
    hasFocus -> {
        focusedActivity = activity
    }
    focusedActivity == activity -> {
        focusedActivity = null
    }
}
```

Could be useful for one-shot listeners but that's a whole another topic.

> Remove the d to make it consistent with the style of ViewTreeObserver interfaces

I'd stay consistent with the method's name on `Window.Callback`, with the d. Don't want to rename OnContentChangedListener, too.

> Remove the interface all together and use the ViewTreeObserver

Can't do, it's API 18+

**Comment by consp1racy** [comment]
Oh but I have another one for you

`RootViewAddedListener` doesn't have `On` at the beginning. Should it?

**Comment by pyricau** [comment]
> I'm tempted to do this 😂

Let's this if that makes it all green, and if yes we can merge and then follow up to clean up if possible.

> I like that, but before I add it, does it make sense?

Yep. Feel free to add or just push the fix and I can follow up.

> RootViewAddedListener doesn't have On at the beginning. Should it?

Great point! I'll get started on a consistency PR

**Comment by consp1racy** [comment]
> Feel free to add or just push the fix

I'll leave it for the API combing

**Comment by pyricau** [comment]
Still failing on API 30. The only possible explanation I can think of is that we paused the activity before it has gained focus.  So we might need an explicit waiting step, see if that helps.

**Comment by consp1racy** [comment]
That's CI, that's a parallel issue, all tests failing on that one. Everything now consistently works locally on API 19-30.

**Comment by pyricau** [comment]
Let me try to push a fix see what happens

File: curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt
Comment (pyricau) [inline curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:109]: I don't know why I went with error() here vs check in the tests. Should probably use check here as well.

Note: I removed the assertThat calls because generally asserts are to be used for the verification part of a test (ie what's being tested) which is different from making sure everything is in the right state for the test to proceed. The result is similar, just different semantics which help understand what the test is about (and yes, probably some other tests need assertThat removal)

Comment (consp1racy) [reply curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:109]: Using the thing under test as a scaffolding around the test, huh :D

> I don't know why I went with error() here vs check in the tests

If it was in the test case, I'd consider `Assert.fail()`. If it's in helper function I'm not sure.

--- 

I had the count downs before assertions because I was afraid of having to wait 30 seconds for the failure to bubble up. The relation was inversed, the latch was just a control mechanism so the activity isn't finished prematurely and then the assertions inside the callback made sense.

How about this instead?
* I use the latch as a control mechanism, same as before, but don't assert on it.
* The listener pushes new items into a `mutableList<Boolean>()`.
* At the absolute end of each test I assert that the list contains the expected value.

Comment (consp1racy) [reply curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt]: Alright, I have it cooked up, let me know when I can blow up the logging

Sneak peek:
obrazek

Basically, don't wait 30 seconds for the exception to bubble up just for getting unexpected `hasFocus` value. Made the same change in the helper. And `check` now makes sense where `error` was.

**Comment by consp1racy** [comment]
If we ran `dumpsys window` through UI Automator shell, could you get the output?

There must be some sort of system dialog or something stealing focus in the CI API 30 emulator, this is ridiculous.

**Comment by pyricau** [comment]
Sorry I just saw your responses.

I had kept the "assert" but migrated them to checks. We don't need to check the counts, because the count won't change between right before we set the listener and right after we set it on the same thread.

I'm not sure we need the queue,  a check has focus on the next call is immediate and good enough, no?

Anyway, I added one more commit: 4499cd8

=> in the helper function if we never get the focus callback it tells us what root views exist:

```
DecorView@9a9fb5f[TestActivity] focus: false callback: curtains.test.utilities.TestActivity@ceea49
```

That means.. no other window is showing, and it's not even that we somehow never got the callback. The window just doesn't get focus??

And yet that doesn't happen on local emulators.

**Comment by pyricau** [comment]
> If we ran dumpsys window through UI Automator shell, could you get the output?
> There must be some sort of system dialog or something stealing focus in the CI API 30 emulator, this is ridiculous.

yeah need to figure out the right config to get more data.

**Comment by pyricau** [comment]
I'm working on setting up the same emulator as CI, see if I can repro.

**Comment by pyricau** [comment]
I ran all the exact commands and still wasn't able to repro. This is a bit frustrating..

**Comment by consp1racy** [comment]
> I'm not sure we need the queue

It combines the latch with passing a value. I'd like to assert the value's right, not just that the callback happened. Latch doesn't quite convey that. 

> the count won't change between right before we set the listener and right after we set it on the same thread

The listener could be made to fire instantly with the current value. But that would break the test anyway so let's not worry about that now.

---

Made some changes, went all fancy with a custom `Subject` for the queue, so there are assertions in the tests. `launchWaitingForFocus` isn't part of the test so it has `check`s, otherwise it's pretty much the same.

Left in the logging so you can pick up later. This is now a WIP. If you're able to grab the window service dump and make it visible in the test results somehow that might help.

I'll continue tomorrow if need be

File: curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt
Comment (pyricau) [inline curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:117]: I don't understand what this changes. Prior to this we had `check(hasFocus) {` so that main thread would throw immediately, effectively making the test fail is that callback was to happen with the wrong boolean. So effectively we have the same situation where we wait 30 seconds if we don't get any callback, and fail immediately if we get a callback with the wrong value.

Not that this is necessarily wrong, and generally this seems like a useful helper, I just don't quite see if / how this helps :)

Comment (consp1racy) [reply curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:117]: If it checked first and threw (so it didn't count down the latch afterwards), then the test thread would still be waiting for the latch, effectively making the failing test always take 30 seconds to complete.

This makes it the test fails near instantly if wrong `hasFocus` comes in.

Comment (pyricau) [reply curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:117]: >  then the test thread would still be waiting for the latch, effectively making the failing test always take 30 seconds to complete.

Oooooh

Comment (consp1racy) [reply curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:117]: Across 12 emulators in the new shiny Android Studio test matrix UI (it's so good)

With `activityWindowFocusChanged.put(!hasFocus)`:

obrazek

With `TODO()`:

obrazek

The exception *on the main thread* happens quickly, yes. But the test still has to wait 30 seconds for the latch/queue. If we deliver the callback value from the main thread *to the test thread*, whatever happens next, the test thread and the test finishes right after that.

**Comment by pyricau** [comment]
I added a commit to not run these tests on API 30. I think also your last commit had a format issue, I'll let you follow up and fix that.

**Comment by consp1racy** [comment]
Too bad we can't check if running in CI from the emulator :(

Thanks, fixed the format issue

**Comment by pyricau** [comment]
Damn, API 29 failed twice.

> Tests on test(AVD) - 10 failed: Test run failed to complete. Expected 41 tests, received 37. onError: commandError=false message=INSTRUMENTATION_ABORTED: System has crashed.

**Comment by consp1racy** [comment]
Oddly enough unrelated

image

`WrappedCallbackTest` only ran 1 out of 5 tests: `wrapped_window_callback_can_be_unwrapped_past_app_compat`

**Comment by consp1racy** [comment]
I squashed it, left the root view dump in the test helper, as it doesn't affect consumers.

I'll branch off this and try the window service dump.

**Comment by consp1racy** [comment]
curtains.test.WrappedCallbackTest > window_callback_not_wrapped_on_new_activity[test(AVD) - 10] FAILED 
Test instrumentation process crashed. Check curtains.test.WrappedCallbackTest#wrapped_window_callback_can_be_unwrapped.txt for details

leftover activity from previous test?

**Comment by consp1racy** [comment]
ANR dialog has focus on API 30 (even more questions). I'll try to send `Intent.ACTION_CLOSE_SYSTEM_DIALOGS` in `@Before`, see if that helps.

obrazek

**Comment by consp1racy** [comment]
Fixed it!

There was an ANR dialog for `com.android.systemui` in the foreground.
A couple of other tests were awaiting the latch in ActivityScenario.onActivity,
which runs on the main thread. I moved the `await` outside, which seems to have helped.

Here's the window service dump as an exception:
https://github.com/consp1racy/curtains/suites/2026422932/artifacts/40687869

Here's the painful process, perhaps there's something handy in there::
https://github.com/consp1racy/curtains/compare/ep/focus...consp1racy:ep/focus-api30-emulator

---

Now there's a new thing on API 19 🙄

**Comment by consp1racy** [comment]
Alright, there's an ANR dialog, which isn't important for resumed activities but plays a role in window focus.

Don't know where the ANR is coming from (apart from `com.android.systemui`). I tried fixing it by moving all latch `await`s in all tests from the main thread, no luck.

We can use `UiDevice` to find the WAIT button in the ANR dialog and press it (API 18+). If that doesn't work we dump the window service (API 21+) and fail. On lower API the test would fail anyway.

**Comment by pyricau** [comment]
Thanks for finding this, wow!

I wonder if the ANR dialog might be caused by something we do? Not sure why, but presumably we're blocking the main thread somewhere and that triggers the ANR?

**Comment by consp1racy** [comment]
It looks like it's from the tests because when I ran just `OnWindowFocusChangedListenersTest` it went fine. But I just ran all the tests locally and didn't reproduce. I don't know.

**Comment by pyricau** [comment]
Sorry it's been too long, I was focusing on other things for a while. So, we know that an ANR is causing this. We know we can dismiss it, but it would be really nice if we could figure out what's causing the ANR, because I don't want to ship code that under some conditions creates ANRs.

File: curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt
Comment (pyricau) [inline curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt:124]: The only thing that comes to mind to explain the ANR would be that we're blocking the main thread. Here I don't think it can happen, from reading the code, but an example would be that ArrayBlockingQueue has a capacity of 1 so if we somehow got 2 calls to put() (i.e. here or somewhere else the listener removal somehow didn't work) then the 2nd put would block. But we're also calling poll so we're removing the entry put anyway. So doesn't seem like it, but I just feel like there's got to be something related to blocking ops on the main thread.

**Comment by consp1racy** [comment]
The code under test is simple. Accommodating this CI emulator is wasting my energy. For all we know the ANR is there since the system started. It only affects window focus tests because it's translucent and other activities can resume under it all day long.

Is there a way to get screenshots or window dump for all tests regardless of the result?

**Comment by pyricau** [comment]
> Accommodating this CI emulator is wasting my energy. 

Yeah thanks for the time you've already put in. I'll try to figure out what's holding up the main thread if anything.

> Is there a way to get screenshots or window dump for all tests regardless of the result?

Probably but I haven't figured that out.

**Comment by consp1racy** [comment]
I'll try just throw the dump (I imagine monkeys throwing 💩 at each other) instead of the test, that way we'll know if the ANR was there before the test began. If so, we can keep dismissing the ANR and tackle it later, separately from the PR.

**Comment by consp1racy** [comment]
All except window focus tests were ignored, the first was `activity_focus_lost_on_dialog_showed`, and this was taken at the very beginning.

obrazek

It took 3 minutes to just start the emulator, perhaps the CI is ill fit for API 30.

**Comment by pyricau** [comment]
> I'll try just throw the dump (I imagine monkeys throwing 💩 at each other) instead of the test, that way we'll know if the ANR was there before the test began. If so, we can keep dismissing the ANR and tackle it later, separately from the PR.

Oh you mean you're going to check if the ANR shows up on master already. Great point!

**Comment by consp1racy** [comment]
Crap, I forgot to link the results: https://github.com/consp1racy/curtains/actions/runs/611896499

Job log: https://github.com/consp1racy/curtains/runs/2007792072?check_suite_focus=true#step:4:411
Test results: https://github.com/consp1racy/curtains/suites/2153193282/artifacts/44106361

**Comment by pyricau** [comment]
> All except window focus tests were ignored, the first was `activity_focus_lost_on_dialog_showed`, and this was taken at the very beginning.
> 
> obrazek
> 
> It took 3 minutes to just start the emulator, perhaps the CI is ill fit for API 30.

Sorry I'm not sure what you mean by "all except window focus tests were ignored".

Are you saying that the ANR dialog shows on start of the CI run?

Also, aren't we having problems with API 29 as well, not just 30?

And overall, are you suggesting that we set up some common code that dismisses any ANR dialog before starting the test suite?

**Comment by pyricau** [comment]
Also:

> Window{42a633c u0 Application Not Responding: com.android.systemui}

Sounds like this is for System UI having an ANR?

**Comment by consp1racy** [comment]
> Are you saying that the ANR dialog shows on start of the CI run?

Yes!

> Also, aren't we having problems with API 29 as well, not just 30?

Yep, but those were unrelated, because of awaiting on the main thread.
See `@Test fun onDecorViewReady_triggers_immediately_if_contentView_set()` and `@Test fun touchEvent_dispatched_to_listeners()` [here](https://github.com/consp1racy/curtains/commit/eda6ed9110818e31a9cd8ebb10cb9fea5e4b5404).

Those should be fixed now. We're stuck only with the pre-existing ANR on API 30 in CI.

> And overall, are you suggesting that we set up some common code that dismisses any ANR dialog before starting the test suite?

That should work around our problem

**Comment by consp1racy** [comment]
> Sounds like this is for System UI having an ANR?

Not enough CPU for the CI job affecting performance inside the emulator?

**Comment by pyricau** [comment]
Ok, you've convinced me! I removed my stupid commits, I totally hadn't realized it was systemui that was having an ANR so really nothing to do with us.

So, to summarize, tell me if all this is correct:

- All tests pass for all API levels, except for API 30 + window focus because of the ANR dialog
- We leverage UI Automator to dismiss that dialog. I think we should probably do that on the window focus test on API 30 since there are no other issues.

Is there any other change you wanted to do? I think there's a TODO left? Let me know where you stand, I have a bit of time right now so I can bring this home but also happy to let you finish.

**Comment by pyricau** [comment]
Filed: https://github.com/ReactiveCircus/android-emulator-runner/issues/129

**Comment by pyricau** [comment]
🎉 thanks for the time you put into this!

**Comment by consp1racy** [comment]
thanks for the opportunity! turned out more interesting

**Pull request merged**
